packages/protocol/contracts/governance/Proposals.sol

```diff
@@ -0,0 +1,330 @@
+    address msgSender,
+    uint256 msgValue
+  )
+    public
+  {
+    require(values.length == destinations.length && destinations.length == dataLengths.length);
+    uint256 transactionCount = values.length;
+
+    proposal.proposer = msgSender;
+    proposal.deposit = msgValue;
+    // solhint-disable-next-line not-rely-on-time
+    proposal.timestamp = now;
+
+    uint256 dataPosition = 0;
+    for (uint256 i = 0; i < transactionCount; i = i.add(1)) {
+      proposal.transactions.push(
+        Transaction(values[i], destinations[i], data.slice(dataPosition, dataLengths[i]))
+      );
+      dataPosition = dataPosition.add(dataLengths[i]);
+    }
+  }
+
+  function vote(
+    Proposal storage proposal,
+    uint256 weight,
+    VoteValue currentVote,
+    VoteValue previousVote
+  )
+    public
+  {
+    // Subtract previous vote.
+    if (previousVote == VoteValue.Abstain) {
+      proposal.votes.abstain = proposal.votes.abstain.sub(weight);
+    } else if (previousVote == VoteValue.Yes) {
+      proposal.votes.yes = proposal.votes.yes.sub(weight);
+    } else if (previousVote == VoteValue.No) {
+      proposal.votes.no = proposal.votes.no.sub(weight);
+    }
+
+    // Add new vote.
+    if (currentVote == VoteValue.Abstain) {
+      proposal.votes.abstain = proposal.votes.abstain.add(weight);
+    } else if (currentVote == VoteValue.Yes) {
+      proposal.votes.yes = proposal.votes.yes.add(weight);
+    } else if (currentVote == VoteValue.No) {
+      proposal.votes.no = proposal.votes.no.add(weight);
+    }
+  }
+
+  /**
+   * @notice Executes the proposal, reverting if any transaction fails.
+   * @param proposal The proposal struct.
+   */
+  function execute(Proposal storage proposal) public {
+    for (uint256 i = 0; i < proposal.transactions.length; i = i.add(1)) {
+      // reverts proposal if any transaction fails
+      require(
+        externalCall(
+          proposal.transactions[i].destination,
+          proposal.transactions[i].value,
+          proposal.transactions[i].data.length,
+          proposal.transactions[i].data
+        )
+      );
+    }
+  }
+
+  /**
+   * @notice Computes the support ratio for a proposal with the quorum condition:
+   *   If the total participation (yes + no + abstain) is less than the critical
+   *   baseline, "no" votes are added to increase particiption to this level. The abstaining votes
+   *   are removed, leaving the "yes" and (potentially increased) "no" votes. The ratio of "yes"
+   *   votes to this value is returned.
+   * @param proposal The proposal struct.
+   * @param criticalBaseline The minimum participation at which "no" votes are not added.
+   * @return The support ratio with the quorum condition.
+   */
+  function adjustedSupport(
+    Proposal storage proposal,
+    int256 criticalBaseline
+  )
+    public
+    view
+    returns (int256)
+  {
+    if (proposal.votes.yes == 0) {
+      return 0;
+    }
+    int256 totalWeightFixed = toFixed(proposal.totalWeight);
```

Comment:
I don't totally understand this function.

Is this psuedo-solidity equivalent?

```
uint256 totalVotes = proposal.votes.yes.add(proposal.votes.no).add(proposal.votes.abstain);
uint256 requiredVotes = criticalBaseline.multiply(proposal.totalWeight);
uint256 yesVotes = proposal.votes.yes;
uint256 noVotes = proposal.votes.no;
if (requiredVotes > totalVotes) {
  noVotes = noVotes.add(requiredVotes.sub(totalVotes));
}
return toFixed(yesVotes).divide(toFixed(yesVotes).add(noVotes);
```

Reply:
That's equivalent. I considered converting to vote counts and computing, but I was concerned about losing precision, so I converted everything to Fixed values instead. Which do you think is better?